File: src/torch_mpc/algos/batch_sampling_mpc.py
import torch
import numpy as np
import matplotlib.pyplot as plt
import time
# import rospy
from torch_mpc.setup_mpc import setup_mpc

class BatchSamplingMPC:
    """
    Implementation of a relatively generic sampling-based MPC algorithm (e.g. CEM, MPPI, etc.)
    We can use this implementation to instantiate a number of algorithms cleanly
    """
    def __init__(self, model, cost_fn, action_sampler, update_rule):
        """
        Args:
            model: The model to optimize over
            cost_fn: The cost function to optimize
            action_sampler: The strategy for sampling actions
            update_rule: The update rule to get best seq from samples
        """
        self.model = model
        self.cost_fn = cost_fn
        self.action_sampler = action_sampler
        self.update_rule = update_rule

        self.device = self.action_sampler.device
        self.B = self.action_sampler.B
        self.H = self.action_sampler.H
        self.K = self.action_sampler.K
        self.n = self.model.observation_space().shape[0] #Note that this won't work for high-dim obses
        self.m = self.model.action_space().shape[0]

        self.setup_variables()

    def setup_variables(self):
        """
        Setup all the variables necessary to run mpc.
        """
        self.last_states = torch.zeros(self.B, self.H, self.n, device=self.device)
        self.last_controls = torch.zeros(self.B, self.H, self.m, device=self.device)

        self.noisy_controls = torch.zeros(self.B, self.K, self.H, self.m, device=self.device)
        self.noisy_states = torch.zeros(self.B, self.K, self.H, self.n, device=self.device)
        self.costs = torch.randn(self.B, self.K, device=self.device)
        self.last_cost = torch.randn(self.B, device=self.device)
        self.last_weights = torch.zeros(self.B, self.K, device=self.device)

        self.u_lb = torch.tensor(self.model.u_lb, device=self.device).view(1, self.m).tile(self.B, 1).float()
        self.u_ub = torch.tensor(self.model.u_ub, device=self.device).view(1, self.m).tile(self.B, 1).float()

    def reset(self):
        self.setup_variables()

    def rollout_model(self,obs,extra_states,noisy_controls):
        states_unsqueeze = torch.stack([obs] * self.K, dim=1)
        if extra_states is not None:
            extra_state_unsqueezed = torch.stack([extra_states] * self.K, dim=1)
            trajs = self.model.rollout(states_unsqueeze, noisy_controls, extra_states=extra_state_unsqueezed)
        else:
            trajs = self.model.rollout(states_unsqueeze, noisy_controls)
        
        return trajs
    
    def get_control(self, obs, step=True, match_noise=False, extra_states=None):
        '''Returns action u at current state obs
        Args:
            x: 
                Current state. Expects tensor of size self.model.state_dim()
            step:
                optionally, don't step the actions if step=False
        Returns:
            u: 
                Action from MPPI controller. Returns Tensor(self.m)
            cost:
                Cost from executing the optimal action sequence. Float.

        Batching convention is [B x K x T x {m/n}], where B = batchdim, K=sampledim, T=timedim, m/n = state/action dim
        '''
        ## Simulate K rollouts and get costs for each rollout
        noisy_controls = self.action_sampler.sample(self.last_controls, self.u_lb, self.u_ub) 
        trajs = self.rollout_model(obs,extra_states,noisy_controls)
        costs, feasible = self.cost_fn.cost(trajs, noisy_controls)

        costs[~feasible] += 1e8 #maybe this is ok?

        optimal_controls, sampling_weights = self.update_rule.update(noisy_controls, costs)

        ## Return first action in sequence and update self.last_controls
        self.last_controls = optimal_controls
        best_cost_idx = torch.argmin(self.costs, dim=1)
        self.last_states = self.noisy_states[torch.arange(self.B), best_cost_idx]
        self.last_cost = self.costs[torch.arange(self.B), best_cost_idx]
        self.noisy_states = trajs
        self.noisy_controls = noisy_controls
        self.costs = costs
        self.last_weights = sampling_weights

        u = self.last_controls[:, 0]

        if step:
            self.step()

        return u, feasible.any(dim=-1)

    def step(self):
        '''Updates self.last_controls to warm start next action sequence.'''
        # Shift controls by one
        self.last_controls = self.last_controls[:, 1:]
        # Initialize last control to be the same as the last in the sequence
        self.last_controls = torch.cat([self.last_controls, self.last_controls[:, [-1]]], dim=1)
#        self.last_controls = torch.cat([self.last_controls, torch.zeros_like(self.last_controls[[-1]])], dim=0)

    def to(self, device):
        self.device = device
        self.u_lb = self.u_lb.to(device)
        self.u_ub = self.u_ub.to(device)
        self.last_controls = self.last_controls.to(device)
        self.noisy_controls = self.noisy_controls.to(device)
        self.noisy_states = self.noisy_states.to(device)
        self.costs = self.costs.to(device)
        self.last_weights = self.last_weights.to(device)

        self.model = self.model.to(device)
        self.cost_fn = self.cost_fn.to(device)
        self.action_sampler = self.action_sampler.to(device)
        return self

    def viz():
        pass

if __name__ == "__main__":
    from torch_mpc.models.skid_steer import SkidSteer
    from torch_mpc.models.kbm import KBM
    from torch_mpc.models.steer_setpoint_throttle_kbm import SteerSetpointThrottleKBM
    from torch_mpc.action_sampling.action_sampler import ActionSampler

    import time
    import yaml

    config_fp = 'C:/Users/anton/Documents/SPRING24/AEC/torch_mpc/configs/costmap_speedmap.yaml'
    config = yaml.safe_load(open(config_fp, 'r'))

    # TODO: integrate w/ config file
    device = config['common']['device']
    batch_size = config['common']['B']

    mppi = setup_mpc(config).to(device)
    model = mppi.model
    cost_fn = mppi.cost_fn
    action_sampler = mppi.action_sampler

    cost_fn.data['goals'] = [
        torch.tensor([
            [5.0, 0.0],
            [10.0, 0.0]
        ]).to(device),
        torch.tensor([
            [3.0, 0.0],
            [4.0, 0.0]
        ]).to(device),
        torch.tensor([
            [6.0, 0.0],
            [4.0, 0.0]
        ]).to(device)
    ]

    cost_fn.data['waypoints'] = [
        torch.tensor([
            [5.0, 0.0],
            [10.0, 0.0]
        ]).to(device),
        torch.tensor([
            [3.0, 0.0],
            [4.0, 0.0]
        ]).to(device),
        torch.tensor([
            [6.0, 0.0],
            [4.0, 0.0]
        ]).to(device)
    ]

    print(cost_fn.can_compute_cost())
    
    cost_fn.data['local_costmap'] = {'metadata': {
        'resolution':torch.tensor([1.0, 0.5, 2.0]),
        'width': torch.tensor([100., 50., 200.]),
        'height': torch.tensor([100., 50., 200.]),
        'origin': torch.tensor([[-50., -50.], [-25., -25.], [-100., -100.]])
    },
                               'data': torch.zeros(3, 100, 100)}
    
    cost_fn.data['costmap']['data'][:, 40:60, 60:] = 10.

    print(cost_fn.can_compute_cost())
    x = torch.zeros(batch_size, model.observation_space().shape[0]).to(device)

    X = []
    U = []

    t0 = time.time()
    for i in range(500):
        X.append(x.clone())
        u = mppi.get_control(x)
        U.append(u.clone())
        x = model.predict(x, u)
    t1 = time.time()

    print('TIME = {:.6f}s'.format(t1 - t0))

    X = torch.stack(X, dim=1).cpu()
    U = torch.stack(U, dim=1).cpu()

    traj = model.rollout(x, mppi.last_controls).cpu()

    print('TRAJ COST = {}'.format(cost_fn.cost(X, U)))

    du = abs(U[:, 1:] - U[:, :-1])
    print('SMOOOTHNESS = {}'.format(du.view(batch_size, -1).mean(dim=-1)))

    t3 = time.time()
    for i in range(100):
        u = mppi.get_control(x)
    t4 = time.time()
    print('ITR TIME = {:.6f}s'.format((t4 - t3)/100.))

    fig, axs = plt.subplots(1, 3, figsize=(18, 6))
    axs[0].set_title("Traj")
    for b in range(batch_size):
        axs[0].plot(X[b, :, 0], X[b, :, 1], c='b', label='traj' if b == 0 else None)
        axs[0].plot(traj[b, :, 0], traj[b, :, 1], c='r', label='pred' if b == 0 else None)
        axs[1].plot(U[b, :, 0], c='r', label='throttle' if b == 0 else None)
        axs[1].plot(U[b, :, 1], c='b', label='steer' if b == 0 else None)

    axs[1].set_title("Controls")
    axs[2].set_title("States")

    colors = 'rgbcmyk'
    for xi, name in enumerate(['X', 'Y', 'Th']):
        c = colors[xi % len(colors)]
        for b in range(batch_size):
            axs[2].plot(X[b, :, xi], c=c, label=name if b == 0 else None)

    axs[0].legend()
    axs[1].legend()
    axs[2].legend()

    # axs[0].set_aspect('equal')
    plt.show()

File: src/torch_mpc/algos/batch_sampling_mpc.cpp
#include "batch_sampling_mpc.h"
#include "../setup_mpc.h"

int main() 
{
    std::string config_file = "config.yaml"; // specify the path to the config file

    YAML::Node config = YAML::LoadFile(config_file);

    // class TestCost
    // {   
    //     private:
    //         torch::Device device;

    //     public:
    //         TestCost()
    //         {
    //             device = torch::kCPU;
    //         }

    //         torch::Tensor cost(torch::Tensor traj, torch::Tensor controls){
    //             auto stage_cost = stage_cost(traj, controls);
    //             auto term_cost = term_cost(traj.index({"...", -1, torch::indexing::Slice()}));
    //             return stage_cost.sum(-1) + term_cost;
    //         }
    //         torch::Tensor stage_cost(const torch::Tensor& traj, const torch::Tensor& controls) {
    //             // drive fast in -x direction
    //             auto state_cost = traj.index({"...", 0}) + 10.0 * (traj.index({"...", 2}) - M_PI).abs();
    //             auto control_cost = torch::zeros_like(state_cost); // Assume control cost is simplified to zero tensor

    //             return state_cost + control_cost;
    //         }
    //         torch::Tensor term_cost(const torch::Tensor& tstates) {
    //             return tstates.index({"...", 1}) * 0.0;
    //         }
    //         TestCost& to(const torch::Device& device) {
    //             this->device = device;
    //             return *this;
    //         }
    // };
 
    const std::string device_config = config["common"]["device"].as<std::string>();
    std::optional<torch::Device> device;
    if (device_config == "cuda")
    {device.emplace(torch::kCUDA, 0);}
    else if (device_config == "cpu")
    {device.emplace(torch::kCPU); }
    else {throw std::runtime_error("Unknown device " + device_config);}

    const int batch_size = config["common"]["B"].as<int>();
    
    // auto cfn = std::make_shared<TestCost>()->to(*device);
    

    // Creating ActionSampler
    // std::unordered_map<std::string, std::unique_ptr<SamplingStrategy>> sampling_strategies;

    // for(auto iter = config["sampling_strategies"]["strategies"].begin(); iter != config["sampling_strategies"]["strategies"].end(); ++iter)
    // {
    //     auto sv = *iter;

    //     std::string type = sv["type"].as<std::string>();
    //     if (type == "UniformGaussian")
    //     {
    //         const int K = sv["args"]["K"].as<int>();
            
    //         const std::vector<double> scale = sv["args"]["scale"].as<std::vector<double>>();
            
    //         auto strategy = std::make_unique<UniformGaussian>(scale, B, K, H, M, *device);
    //         sampling_strategies.emplace(sv["label"].as<std::string>(), std::move(strategy));
    //     }
    //     else if (type == "ActionLibrary")
    //     {
    //         const int K = sv["args"]["K"].as<int>();
            
    //         std::string path = sv["args"]["path"].as<std::string>();
        
    //         auto strategy = std::make_unique<ActionLibrary>(path, B, K, H, M, *device);
    //         sampling_strategies.emplace(sv["label"].as<std::string>(), std::move(strategy));
    //     }
    //     else if (type == "GaussianWalk")
    //     {
    //         const int K = sv["args"]["K"].as<int>();
            
    //         const std::vector<double> scale = sv["args"]["scale"].as<std::vector<double>>();
        
    //         std::unordered_map<std::string, std::variant<std::string, std::vector<double>>> initial_distribution;

    //         initial_distribution["type"] = sv["args"]["initial_distribution"]["type"].as<std::string>();
    //         initial_distribution["scale"] = sv["args"]["initial_distribution"]["scale"].as<std::vector<double>>();

    //         const std::vector<double> alpha = sv["args"]["alpha"].as<std::vector<double>>();

    //         auto strategy = std::make_unique<GaussianWalk>(initial_distribution, scale, alpha, B, K, H, M, *device);
    //         sampling_strategies.emplace(sv["label"].as<std::string>(), std::move(strategy));
            
    //     }
    //     else
    //     {
    //         throw std::runtime_error("Unknown sampling strategy " + type);
    //     }
    // }

    // // ActionSampler action_sampler(sampling_strategies);
    // auto action_sampler = std::make_shared<ActionSampler>(sampling_strategies);

    // Creating model
    // auto model = std::make_shared<KBM>(3.0, 0., 1., 0.3, 0.1, *device); 

    // // Creating update rule
    // auto update_rule = std::make_shared<MPPI>(config["update_rule"]["args"]["temperature"].as<double>());
    
    // Creating MPC
    // auto mppi = std::make_unique<BatchSamplingMPC>(model, cfn, mppi, action_sampler, update_rule);

    auto mppi = setup_mpc(config); // returns a unique pointer to BatchSamplingMPC
    auto model = mppi->model; // returns a shared pointer to Model
    auto cfn = mppi->cost_function; // returns a shared pointer to CostFunction

    auto x = torch::zeros({batch_size, model->observation_space()}, torch::Options(*device));

    std::vector<torch::Tensor> X; // X is the state
    std::vector<torch::Tensor> U; // U is the control input

    auto t0 = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 500; i++)
    {
        X.push_back(x.clone());
        auto u = mppi->get_control(x);
        U.push_back(u.clone());
        x = model->predict(x, u);
    }
    auto t1 = std::chrono::high_resolution_clock::now();

    std::cout << "TIME: " << std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0).count() << " milliseconds" << std::endl;

    auto X_tensor = torch::stack(X, 1).to(torch::kCPU);
    auto U_tensor = torch::stack(U, 1).to(torch::kCPU);

    auto traj = model->rollout(X_tensor, mppi->last_controls).to(torch::kCPU);

    std::cout << "TRAJ COST = " << cfn->cost(X_tensor, U_tensor) << std::endl;

    auto du = torch::abs(U_tensor.slice(1,1) - U_tensor.slice(1,0,-1));

    std::cout << "SMOOTHNESS = " << du.view({batch_size, -1})mean(-1) << std::endl;

    auto t3 = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < 100; i++)
    {
        auto u = mppi->get_control(x);
    }
    auto t4 = std::chrono::high_resolution_clock::now();
    std::cout << "TIME: " << std::chrono::duration_cast<std::chrono::milliseconds>(t4 - t3).count() << " milliseconds" << std::endl;

    std::string dir_path = "/home/anton/Desktop/SPRING24/AEC/torch_mpc/src/torch_mpc/algos/algos_data/";
    torch::save(X_tensor, dir_path + "X.pt");
    torch::save(U_tensor, dir_path + "U.pt");
    torch::save(traj, dir_path + "traj.pt");

    std::cout << "Saved data" << std::endl;

    return 0;
}
